`CRoaring-4.7.0` fails to build when forcing old cmocka (ROARING_USE_CPM=OFF + system)

**Describe the bug**
When building croaring v4.7.0 with tests enabled and a modern C23 compiler (with C23 changes `-Wimplicit-function-declaration` is now an error instead of a warning only), croaring fails to build
```sh
[27/72] /usr/bin/gcc-15  -I/tmp/CRoaring-4.7.0/include -I/tmp/CRoaring-4.7.0/cpp -Wall -Wextra \
-Wsign-compare -Wshadow -Wwrite-strings -Wpointer-arith -Winit-self -Wcast-align  -O2 -g -DNDEBUG \
-std=gnu11 -MD -MT tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o -MF tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o.d -o tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o \
-c /tmp/CRoaring-4.7.0/tests/toplevel_unit.c
FAILED: [code=1] tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o 
/usr/bin/gcc-15  -I/tmp/CRoaring-4.7.0/include -I/tmp/CRoaring-4.7.0/cpp -Wall -Wextra \
-Wsign-compare -Wshadow -Wwrite-strings -Wpointer-arith -Winit-self -Wcast-align  -O2 -g -DNDEBUG \
-std=gnu11 -MD -MT tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o -MF tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o.d -o tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o \
-c /tmp/CRoaring-4.7.0/tests/toplevel_unit.c
/tmp/CRoaring-4.7.0/tests/toplevel_unit.c: In function ‘contains_bulk’:
/tmp/CRoaring-4.7.0/tests/toplevel_unit.c:202:13: error: implicit declaration of function ‘assert_uint_in_range’; did you mean ‘assert_not_in_range’? [-Wimplicit-function-declaration]
  202 |             assert_uint_in_range(context.idx, 0,
      |             ^~~~~~~~~~~~~~~~~~~~
      |             assert_not_in_range
```
**To Reproduce**
- `wget https://github.com/RoaringBitmap/CRoaring/archive/refs/tags/v4.7.0.tar.gz`
- `tar -xf v4.7.0.tar.gz`
- `cd CRoaring-4.7.0/`
- `export CC=gcc-15 CXX=g++-15`
- `cmake -G Ninja -DBUILD_SHARED_LIBS=OFF -DENABLE_ROARING_TESTS=ON -DROARING_USE_CPM=OFF -DCMAKE_BUILD_TYPE=RelWithDebInfo`
- `ninja -v -j9`
- See the `[-Wimplicit-function-declaration]` failure when buildiing


**Configuration (please complete the following information if relevant)**
 - OS: [Gentoo linux] (I am packaging croaring)
 - Compiler* [Fails with GCC-15, GCC-16 and Clang-21 (x86_64-pc-linux-gnu)] (the implicite fucntion declaration became an error with >=gcc-14)
 - Optimization setting (-O2 -g)

We support up-to-date 64-bit ARM and x64 FreeBSD, macOS, Windows and Linux systems. Please ensure that your configuration is supported before labelling the issue as a bug.

**Indicate whether you are willing or able to provide a bug fix as a pull request**

I don't know C so I won't be able to provide a patch.

If I pass `-Wno-implicit-function-declaration` to `CFLAGS` I then get:
```sh
FAILED: [code=1] tests/toplevel_unit 
: && /usr/bin/gcc-15 -O2 -g -Wno-implicit-function-declaration    -Wall -Wextra \
-Wsign-compare -Wshadow -Wwrite-strings -Wpointer-arith -Winit-self -Wcast-align -Wl,--as-needed,-O1 -fuse-ld=bfd \
-Wl,--dependency-file=tests/CMakeFiles/toplevel_unit.dir/link.d tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o -o tests/toplevel_unit  \
-Wl,-rpath,/var/tmp/portage/dev-libs/croaring-4.7.0/work/CRoaring-4.7.0_build  libroaring.so.4.7.0  -lcmocka && :
/usr/x86_64-pc-linux-gnu/binutils-bin/2.46.0/ld.bfd: tests/CMakeFiles/toplevel_unit.dir/toplevel_unit.c.o: in function `contains_bulk':
/var/tmp/portage/dev-libs/croaring-4.7.0/work/CRoaring-4.7.0/tests/toplevel_unit.c:202:(.text+0x7cb2): undefined reference to `assert_uint_in_range'
collect2: error: ld returned 1 exit status
```
So it looks like a real bug.

--- a/.github/workflows/ubuntu-c23-ci.yml
+++ b/.github/workflows/ubuntu-c23-ci.yml
@@ -1,0 +1,36 @@
+name: Ubuntu-C23-CI
+
+'on':
+  - push
+  - pull_request
+
+permissions:
+  contents: read
+
+jobs:
+  ci:
+    name: ubuntu-c23-${{ matrix.compiler.cc }}
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: false
+      matrix:
+        compiler:
+          - { cc: gcc, cxx: g++ }
+          - { cc: clang, cxx: clang++ }
+
+    env:
+      CC: ${{ matrix.compiler.cc }}
+      CXX: ${{ matrix.compiler.cxx }}
+
+    steps:
+      - uses: actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8 # v6.0.1
+      - name: Build and Test (C23 / C++23)
+        run: |
+          mkdir build
+          cd build
+          cmake -DCMAKE_C_STANDARD=23 -DCMAKE_CXX_STANDARD=23 -DCMAKE_INSTALL_PREFIX:PATH=destination -DENABLE_ROARING_TESTS=ON -DROARING_UNSAFE_FROZEN_TESTS=ON ..
+          cmake --build .
+          ctest . --output-on-failure
+          cmake --install .
+          cd ../tests/installation/find && mkdir build && cd build && cmake -DCMAKE_INSTALL_PREFIX:PATH=../../../build/destination .. && cmake --build .

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3,8 +3,12 @@ project(RoaringBitmap
   DESCRIPTION "Roaring bitmaps in C (and C++)"
   LANGUAGES CXX C
 )
-set (CMAKE_C_STANDARD 11)
-set (CMAKE_CXX_STANDARD 11)
+if (NOT DEFINED CMAKE_C_STANDARD)
+  set (CMAKE_C_STANDARD 11)
+endif()
+if (NOT DEFINED CMAKE_CXX_STANDARD)
+  set (CMAKE_CXX_STANDARD 11)
+endif()
 
 include(GNUInstallDirs)
 

--- a/tests/test.h
+++ b/tests/test.h
@@ -57,6 +57,11 @@ extern "C" {
 //
 #ifdef __cplusplus
 #define DEFINE_TEST(name) static void name(void **)
+#elif defined(__STDC_VERSION__) && __STDC_VERSION__ >= 202311L
+// In C23, `void name()` declares a function taking no arguments, which is
+// incompatible with cmocka's `void (*)(void **)`. C23 lets us name the
+// parameter and mark it unused to keep the cmocka-compatible signature.
+#define DEFINE_TEST(name) static void name([[maybe_unused]] void **state)
 #else
 #define DEFINE_TEST(name) static void name()
 #endif

--- a/tools/cmake/FindCTargets.cmake
+++ b/tools/cmake/FindCTargets.cmake
@@ -9,7 +9,7 @@ if (ENABLE_ROARING_TESTS AND NOT EMSCRIPTEN)
       OPTIONS "UNIT_TESTING ON"
     )
   else()
-    find_package(cmocka REQUIRED)
+    find_package(cmocka 2.0.0 REQUIRED)
   endif()
 endif()
 

